src/igraph_eulerian.c

```diff
@@ -0,0 +1,612 @@
+
+    e_count = igraph_ecount(graph);
+    IGRAPH_CHECK(igraph_vector_bool_init(&visited_list, e_count));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &visited_list);
+
+    for (i = 0; i < e_count; i++) {
+        VECTOR(visited_list)[i] = 0;
+    }
+
+    IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+
+    IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+    for (i = 0; i < igraph_vector_size(&degree); i++) {
+        if (((long int) VECTOR(degree)[i]) > 0) {
+            start = i;
+            break;
+        }
+    }
+
+    igraph_vector_destroy(&degree);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &incl, IGRAPH_ALL));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &incl);
+
+    IGRAPH_CHECK(igraph_vector_init(&bridges, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &bridges);
+    IGRAPH_CHECK(igraph_bridges(graph, &bridges));
+
+    if (cycle && !has_path) {
+        IGRAPH_CHECK(euler_undirected_implementation(start, graph, path, &visited_list, &bridges));
+    } else {
+        for (i = 0; i < igraph_vcount(graph); i++) {
+            incedges = igraph_inclist_get(&incl, i);
+            if ( igraph_vector_int_size(incedges) % 2 == 1) {
+                start = i;
+                break;
+            }
+        }
+        euler_undirected_implementation(start, graph, path, &visited_list, &bridges);
+    }
+
+    igraph_inclist_destroy(&incl);
+    igraph_vector_destroy(&bridges);
+    igraph_vector_bool_destroy(&visited_list);
+    IGRAPH_FINALLY_CLEAN(3);
+
+    return IGRAPH_SUCCESS;
+
+}
+
+/* solution adapted from https://www.geeksforgeeks.org/hierholzers-algorithm-directed-graph/ */
+int eulerian_path_directed_implementation(igraph_t *graph, igraph_integer_t *start_node, igraph_vector_t *outgoing_list, igraph_vector_t *res) {
+    igraph_integer_t start = *start_node;
+    igraph_integer_t curr = start;
+    igraph_integer_t next, e_count, curr_e;
+    igraph_inclist_t il;
+    igraph_stack_t path, tracker, edge_tracker, edge_path;
+    igraph_vector_int_t *incedges;
+    igraph_vector_bool_t visited_list;
+    long nc, edge;
+    int i, j;
+
+    IGRAPH_CHECK(igraph_stack_init(&path, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &path);
+
+    IGRAPH_CHECK(igraph_stack_init(&tracker, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &tracker);
+
+    IGRAPH_CHECK(igraph_stack_init(&edge_path, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &edge_path);
+
+    IGRAPH_CHECK(igraph_stack_init(&edge_tracker, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &edge_tracker);
+
+    e_count = igraph_ecount(graph);
+    IGRAPH_CHECK(igraph_vector_bool_init(&visited_list, e_count));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &visited_list);
+
+    for (i = 0; i < e_count; i++) {
+        VECTOR(visited_list)[i] = 0;
+    }
```

Comment:
Default initialization might have zero everywhere already? just a little note

Reply:
Even if it is not, we have `igraph_vector_fill()`. But I think the default is to have zeros everywhere as we use `calloc()` for memory allocation.